source/fuzz/transformation_invert_comparison_operator.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2020 Vasyl Teliman
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_invert_comparison_operator.h"
+
+#include <utility>
+
+#include "source/fuzz/fuzzer_util.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationInvertComparisonOperator::TransformationInvertComparisonOperator(
+    protobufs::TransformationInvertComparisonOperator message)
+    : message_(std::move(message)) {}
+
+TransformationInvertComparisonOperator::TransformationInvertComparisonOperator(
+    uint32_t operator_id, uint32_t fresh_id) {
+  message_.set_operator_id(operator_id);
+  message_.set_fresh_id(fresh_id);
+}
+
+bool TransformationInvertComparisonOperator::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // |message_.operator_id| must be valid and inversion must be supported for
+  // it.
+  auto* inst = ir_context->get_def_use_mgr()->GetDef(message_.operator_id());
+  if (!inst || !IsInversionSupported(inst->opcode())) {
+    return false;
+  }
+
+  // Check that we can insert negation instruction.
+  auto* block = ir_context->get_instr_block(inst);
+  assert(block && "Instruction must have a basic block");
+
+  auto iter = fuzzerutil::GetIteratorForInstruction(block, inst);
+  ++iter;
+  assert(iter != block->end() && "Instruction can't be the last in the block");
+  assert(fuzzerutil::CanInsertOpcodeBeforeInstruction(SpvOpLogicalNot, iter) &&
+         "Can't insert negation after comparison operator");
+
+  // |message_.fresh_id| must be fresh.
+  return fuzzerutil::IsFreshId(ir_context, message_.fresh_id());
+}
+
+void TransformationInvertComparisonOperator::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto* inst = ir_context->get_def_use_mgr()->GetDef(message_.operator_id());
+  assert(inst);
```

Comment:
I like to have an ' && "message"' with every assertion.

Reply:
Fixed in 682be74.